connector/keystone/keystone_test.go

```diff
@@ -0,0 +1,326 @@
+  	if !validPW {
+     	t.Fail()
+  	}
+  	delete(token, userID, usersURL)
+}
+
+func TestUseRefreshToken(t *testing.T) {
+	token, adminID := getAdminToken(adminUser, adminPass)
+	groupID := createGroup(token, "Test group description", testGroup)
+	addUserToGroup(token, groupID, adminID)
+
+	c := keystoneConnector{KeystoneHost: keystoneURL, Domain: domain,
+		KeystoneUsername: adminUser, KeystonePassword: adminPass}
+	s := connector.Scopes{OfflineAccess: true, Groups: true}
+
+	identityLogin, _, _ := c.Login(context.Background(), s, adminUser, adminPass)
+	identityRefresh, _ := c.Refresh(context.Background(), s, identityLogin)
+
+	delete(token, groupID, groupsURL)
+
+	assert.Equal(t, 1, len(identityRefresh.Groups))
+	assert.Equal(t, testGroup, string(identityRefresh.Groups[0]))
+}
+
+func TestUseRefreshTokenUserDeleted(t *testing.T){
+	token, _ := getAdminToken(adminUser, adminPass)
+	userID := createUser(token, testUser, testEmail, testPass)
+
+	c := keystoneConnector{KeystoneHost: keystoneURL, Domain: domain,
+		KeystoneUsername: adminUser, KeystonePassword: adminPass}
+	s := connector.Scopes{OfflineAccess: true, Groups: true}
+
+	identityLogin, _, _ := c.Login(context.Background(), s, testUser, testPass)
+	c.Refresh(context.Background(), s, identityLogin)
+
+	delete(token, userID, usersURL)
+	_, response := c.Refresh(context.Background(), s, identityLogin)
+
+	assert.Contains(t, response.Error(), "does not exist")
+}
+
+func TestUseRefreshTokenGroupsChanged(t *testing.T){
+	token, _ := getAdminToken(adminUser, adminPass)
+	userID := createUser(token, testUser, testEmail, testPass)
+
+	c := keystoneConnector{KeystoneHost: keystoneURL, Domain: domain,
+		KeystoneUsername: adminUser, KeystonePassword: adminPass}
+	s := connector.Scopes{OfflineAccess: true, Groups: true}
+
+	identityLogin, _, _ := c.Login(context.Background(), s, testUser, testPass)
+	identityRefresh, _ := c.Refresh(context.Background(), s, identityLogin)
+
+	assert.Equal(t, 0, len(identityRefresh.Groups))
+
+	groupID := createGroup(token, "Test group description", testGroup)
+	addUserToGroup(token, groupID, userID)
+
+	identityRefresh, _ = c.Refresh(context.Background(), s, identityLogin)
+
+	delete(token, groupID, groupsURL)
+	delete(token, userID, usersURL)
+
+	assert.Equal(t, 1, len(identityRefresh.Groups))
+}
+
+func TestMain(m *testing.M) {
+	testKeystoneEnv := "DEX_TEST_KEYSTONE"
+	endpointsStr := os.Getenv(testKeystoneEnv)
+	if endpointsStr == "" {
+		fmt.Printf("variable %q not set, skipping keystone connector tests\n", testKeystoneEnv)
+		return
+	}
+	dockerID := startKeystoneContainer()
+  	repeats := 10
+  	running := false
+  	for i := 0; i < repeats; i++ {
+   		_, err := http.Get(keystoneURL)
+   		if err == nil {
+     		running = true
+     		break
+   		}
+   		time.Sleep(10 * time.Second)
+  	}
+  	if !running {
+    	fmt.Printf("Failed to start keystone container")
+    	os.Exit(1)
+  	}
+  	defer cleanKeystoneContainer(dockerID)
+  	// run all tests
+	m.Run()
```

Comment:
❓ Do we need `os.Exit(m.Run())` here to have the test run fail when one of the tests fails?